File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/Jackson.java
package uk.nhs.nhsx.core;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;

import java.io.IOException;
import java.io.InputStream;
import java.util.Optional;
import java.util.function.Consumer;

public class Jackson {

    public static <T> T readJson(String value, Class<T> clazz) throws JsonProcessingException {
        return SystemObjectMapper.MAPPER.readValue(value, clazz);
    }

    public static <T> T readJson(InputStream inputStream, Class<T> clazz) throws IOException {
        return SystemObjectMapper.MAPPER.readValue(inputStream, clazz);
    }

    public static <T> T readJson(String value, TypeReference<T> clazz) throws IOException {
        return SystemObjectMapper.MAPPER.readValue(value, clazz);
    }

    public static <T> T readJson(InputStream inputStream, TypeReference<T> clazz) throws IOException {
        return SystemObjectMapper.MAPPER.readValue(inputStream, clazz);
    }

    public static <T> Optional<T> readMaybe(String value, Class<T> toClass, Consumer<Exception> handleError) {
        return Optional.ofNullable(readOrNull(value, toClass, handleError));
    }

    public static <T> T readOrNull(String value, Class<T> toClass, Consumer<Exception> handleError) {
        try {
            return readJson(value, toClass);
        } catch (Exception e) {
            handleError.accept(e);
            return null;
        }
    }

    public static String toJson(Object value) {
        try {
            return SystemObjectMapper.MAPPER.writeValueAsString(value);
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/Jackson.kt
package uk.nhs.nhsx.core

import com.fasterxml.jackson.core.JsonProcessingException
import uk.nhs.nhsx.core.SystemObjectMapper.MAPPER
import uk.nhs.nhsx.core.SystemObjectMapper.STRICT_MAPPER
import java.io.InputStream
import java.util.function.Consumer

object Jackson {

    fun <T> readJsonOrThrow(inputStream: InputStream?, clazz: Class<T>): T = MAPPER.readValue(inputStream, clazz)

    inline fun <reified T> readJsonOrThrow(value: String?): T = MAPPER.readValue(value, T::class.java)

    inline fun <reified T> readOrNull(value: String?, handleError: Consumer<Exception> = Consumer { }): T? = try {
        readJsonOrThrow<T>(value)
    } catch (e: Exception) {
        handleError.accept(e)
        null
    }

    inline fun <reified T> readStrictOrNull(value: String?, handleError: Consumer<Exception> = Consumer { }): T? = try {
        STRICT_MAPPER.readValue(value, T::class.java)
    } catch (e: Exception) {
        handleError.accept(e)
        null
    }

    fun toJson(value: Any?): String = try {
        MAPPER.writeValueAsString(value)
    } catch (e: JsonProcessingException) {
        throw RuntimeException(e)
    }
}
